certification/artifacts/filesystem_writer.go

```diff
@@ -0,0 +1,58 @@
+package artifacts
+
+import (
+	"fmt"
+	"io"
+	"path/filepath"
+
+	"github.com/spf13/afero"
+)
+
+// FilesystemWriter is an ArtifactWriter that targets a particular directory on
+// the underlying filesystem.
+type FilesystemWriter struct {
+	dir string
+	fs  afero.Fs
+}
+
+// NewFilesystemWriter creates an artifact writer which writes to the filesystem.
+func NewFilesystemWriter(opts ...FilesystemWriterOption) (*FilesystemWriter, error) {
+	w := FilesystemWriter{
+		dir: resolveFullPath(DefaultArtifactsDir),
+		fs:  afero.NewOsFs(),
+	}
+
+	for _, opt := range opts {
+		opt(&w)
+	}
+
+	return &w, nil
+}
+
+// WithDirectory sets the artifacts directory to dir unless it's empty, in which case
+// this option is ignored.
+func WithDirectory(dir string) FilesystemWriterOption {
+	return func(w *FilesystemWriter) {
+		if dir == "" {
+			return
+		}
+		w.dir = resolveFullPath(dir)
+	}
+}
+
+type FilesystemWriterOption = func(*FilesystemWriter)
```

Comment:
Discussed this with you because I could go either way. I'm going to go with no error here, simply because I don't want to inadvertently encourage the possibility of doing option evaluations here, and returning an error if there are two exclusive options. Option ordering would cause this to be flaky.

To your point, the intended use of an error in an option is for things that are invalid (nil values to interface parameters, etc). That **IS** a valid use case that will need to be handled prior internally before any logic is called. You have an example of this in another project, so having no error returned here will serve as a bit of an A/B test.